make 'Select Dupes' actually select dupes

szero opened the pull request.

szero committed:
make 'Select Dupes' actually select dupes

szero committed:
bump version

RealDolos commented on ownerdelete.user.js:
```diff
@@ -163,6 +163,16 @@ dry.once("load", () => {
     }
     return file;
   };
+  async function getInfo(id) {
```
I REALLY don't like that you use "let obj" i.e. a variable at all.
Just return the result or return null in the catch

RealDolos commented on ownerdelete.user.js:
```diff
@@ -217,10 +227,13 @@ dry.once("load", () => {
         {white: true} :
         {green: true});
       file.dom.setTags(tags);
+      getInfo(file.id).then(result => {
```
Lain will hate you a) for getting the info ALWAYS and b) for getting all infos at once basically. (A 1000 file room would dispatch 1000 getinfo requests at once).

You should probably a) add a cache<sup>1</sup>, that is somewhat persistent but not too persistent (e.g. sessionStorage) and b) use something (like promisepool as volanail does) to limit the number of concurrent in-flight requests.

<sup>1</sup> something I considered to do for volanail but it's a little harder there as the information nail uses may be updated later, while the checksum will not update so that's not an issue here

RealDolos commented on ownerdelete.user.js:
```diff
@@ -310,10 +323,10 @@ dry.once("load", () => {
         const known = new Set();
         dry.exts.filelistManager.filelist.filelist.forEach(e => {
-          const k = `${e.size}/${e.name}`;
+          const k = e.checksum;
```
The checksum might not be available yet, or at all, at the getinfo call might not have returned or errored out.
At the very least, you gotta check for that.
Better yet, you'd store in e.checksum the actual getinfo promise, so you can wait for it (and display some notice that the list is still being populated if it takes too long)

RealDolos replied:
PS: if checksum is null or undefined, you' potentially select a bunch of unrelated files (all having null/undefined) which aren't actual dupes, so this is a major bug.

szero replied:
I added early if case that check if promise pool responsible for assigning checksums to files is still processing. If not, than we can select dupes.

RealDolos replied:
STILL should make sure the API actually returned something sensible

szero committed:
limit concurrency and add sessionStorage cache for checsums

'Select Dupes' only works when PromisePool isn't processing anything

szero committed:
remove stupid iife

RealDolos commented on ownerdelete.user.js:
```diff
@@ -141,7 +143,12 @@ dry.once("dom", () => {
   let last_file = null;
   const ownerFiles = new WeakMap();
+  const pool = new PromisePool(6);
+  const checksums = JSON.parse(sessionStorage.getItem("ownerChecksums")) || {};
```
Use Map

RealDolos commented on ownerdelete.user.js:
```diff
@@ -141,7 +143,12 @@ dry.once("dom", () => {
+  const checksums = JSON.parse(sessionStorage.getItem("ownerChecksums")) || {};
 
+  const save_checksums = function() {
+    sessionStorage.setItem("ownerChecksums", JSON.stringify(checksums));
```
Also Map. Array.from(map) is your friend ;)

RealDolos commented on ownerdelete.user.js:
```diff
@@ -163,6 +170,22 @@ dry.once("load", () => {
+      const {checksum} = info;
+      checksums[file.id] = checksum;
+      file.checksum = checksum;
+      save_checksums();
```
This shouldn't be in the hot path. 1000 files -> 1000 session storage saves is excessive.
E.g. you could start a timer, if there isn't one yet, to save once after a certain amount of time to effectively coalesce saves.

RealDolos commented on ownerdelete.user.js:
```diff
@@ -163,6 +170,22 @@ dry.once("load", () => {
+        file.checksum = checksums[file.id];
+        return;
+      }
+      const info = await dry.exts.info.getFileInfo(file.id);
```
Really, this line is basically the only line you actually want to run in the promise pool. Well, and some error handling maybe. That way you can avoid going through the microtask queue when the value is already cached.

Actually, I ran into a bug in getFileInfo in volanail that might happen here too: if a file vanishes between queuing on the pool and the pool actually running getFileInfo, then the promise getFileInfo may never resolve, which in turn would block the promise pool too, of course. (lain cannot code).
I didn't address this in volanail yet, tho, as it didn't seem worth spending time on. Basically, a potential fix would be to make a promise that gets resolved by getFileInfo but rejected after a timeout.

RealDolos commented on ownerdelete.user.js:
```diff
@@ -217,10 +240,10 @@ dry.once("load", () => {
         {white: true} :
         {green: true});
       file.dom.setTags(tags);
+      pool.schedule(getInfo, file);
```
I personally prefer pool.wrap (aka bind a function to always use  the pool) instead of explicitly scheduling every call.

szero replied:
I use explicit calls because when I wrap the function, I don't have access to the getters, which are required later.

RealDolos commented on ownerdelete.user.js:
```diff
@@ -364,6 +391,12 @@ dry.once("load", () => {
     dry.exts.filelistManager.on("fileAdded", prepare_file);
     dry.exts.filelistManager.on("fileUpdated", prepare_file);
+    dry.exts.filelistManager.on("fileRemoved", file => {
+      if (file.id in checksums) {
```
Map

RealDolos commented on ownerdelete.user.js:
```diff
@@ -163,6 +170,22 @@ dry.once("load", () => {
   };
+  async function getInfo(file) {
+    try {
+      if (file.id in checksums && !file.checksum) {
```
Map

Also, if file.checksum is true, shouldn't it early-return too?

RealDolos commented on ownerdelete.user.js:
```diff
@@ -364,6 +391,12 @@ dry.once("load", () => {
+    dry.exts.filelistManager.on("fileRemoved", file => {
+      if (file.id in checksums) {
+        delete checksums[file.id];
+        save_checksums();
```
potentially excessive save_checksums, e.g. if a mod nukes a room. Same as above.

szero committed:
replace Object with Map

add debouncing for saving the saving the checksums
fix some possible esoteric bug with getFileInfo by adding a timeout
rejecting promise that races against getFileInfo
retry acquiring fileinfo in `Select Dupes` if its stil not set

RealDolos commented on ownerdelete.user.js:
```diff
@@ -308,12 +372,21 @@ dry.once("load", () => {
+          const k = e.checksum;
+          if (!k) {
+            // reschedule failed item and skip to the next iteration
+            pool.schedule(getInfo, e);
```
this will loop when a file expired and the getter sets it to false. So no

RealDolos commented on ownerdelete.user.js:
```diff
@@ -364,6 +437,11 @@ dry.once("load", () => {
     dry.exts.filelistManager.on("fileUpdated", prepare_file);
+    dry.exts.filelistManager.on("fileRemoved", file => {
+      if (checksums.has(file.id)) {
+        checksums.delete(file.id);
```
Just delete, no need to has first.

RealDolos commented on ownerdelete.user.js:
```diff
@@ -141,7 +171,21 @@ dry.once("dom", () => {
+    let rv = sessionStorage.getItem("ownerChecksums");
+    rv = JSON.parse(rv || "[]");
+    try {
+      return new Map(rv);
```
rv && JSON.parse(rv) should work...

szero committed:
remove "has" from file deletion

simplify session saving

remove reshedule for checksum
